File: Algorithms-and-data-structures-master/0.3 Is binary tree search/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <limits>
#include <string>

struct Node {
    long long value;
    int left = -1;
    int right = -1;
    int parent = -1;
};

std::string is_binary_tree_search(const std::vector<Node>& tree, int nodes_amount) {


    std::vector<std::pair<long long, long long>> intervals;
    intervals.push_back({ LLONG_MIN, LLONG_MAX });

    for (int i = 1; i < nodes_amount; i++) {
        long long value = tree[i].value;
        int parent_index = tree[i].parent;
        long long min_value, max_value;

        if (tree[parent_index].left == i) {
            min_value = intervals[parent_index].first;
            max_value = tree[parent_index].value;
        }
        else {
            min_value = tree[parent_index].value;
            max_value = intervals[parent_index].second;
        }

        intervals.push_back({ min_value, max_value });
        if (value < min_value || value >= max_value)
            return "NO";
    }
    return "YES";
}

int main() {
    std::ifstream fin("bst.in");
    std::ofstream fout("bst.out");

    int n;
    fin >> n;

    if (n == 0) {
        fout << "YES";
        return 0;
    }

    std::vector<Node> tree(n);
    fin >> tree[0].value;

    for (int i = 1; i < n; i++) {
        long long value;
        int parent;
        char side;
        fin >> value >> parent >> side;
        tree[i].value = value;
        tree[i].parent = parent - 1;

        if (side == 'L') {
            tree[parent - 1].left = i;
        }
        else {
            tree[parent - 1].right = i;
        }
    }

    fout << is_binary_tree_search(tree, n);

    return 0;
}

File: Algorithms-and-data-structures-master/0.3 Is binary tree search/main.py


def is_binary_tree_search(input_data, nodes_amount):
    list_of_intervals = []
    list_of_intervals.append((-float('Inf'), float('Inf')))
    for i in range(1, nodes_amount):
        value, parent_index, side = int(input_data[i][0]), int(input_data[i][1]), input_data[i][2]
        if side == "L":
            min_value = list_of_intervals[parent_index-1][0]
            max_value = int(input_data[parent_index-1][0])
        elif side == "R":
            min_value = int(input_data[parent_index-1][0])
            max_value = list_of_intervals[parent_index-1][1]
        list_of_intervals.append((min_value, max_value))
        if (value < min_value or value >= max_value):
            return "NO"

    return "YES"


input_data = []
with open("bst.in", "r") as f:
    nodes_amount = int(f.readline())
    input_data.append([f.readline().strip(), 0, ""])
    for i in range(nodes_amount - 1):
        input_data.append(f.readline().split())

with open("bst.out", "w") as f:
    f.write(is_binary_tree_search(input_data, nodes_amount))
